reana_server/rest/launch.py

```diff
@@ -198,7 +206,18 @@ def launch(user, url, name="", parameters="{}", specification=None):
             500,
         )
     finally:
-        remove_fetched_workflows_dir(tmpdir)
+        # FIXME: `load_reana_spec` is not thread-safe and it changes the cwd, so for the
+        # time being we can only delete the files inside the directory where the
+        # workflow is fetched.  Deleting the directory would result in a
+        # `FileNotFoundError` when calling `os.getcwd()`, due to the cwd not existing
+        # anymore.
+        #
+        # remove_fetched_workflows_dir(tmpdir)
+        for entry in os.scandir(tmpdir):
```

Comment:
I guess it should be OK to iterate through everything and delete one by one, since it is probable that there will be only a few spec files fetched there... But I'm wondering out loud whether this may become slow in certain situations (shared disk busy!) which could contribute to endpoint responsiveness troubles.

Perhaps we can make a length check to see how many files would be deleted, and do the deletions "live" only when there are less than N files? Or delete via one command from the root, such as `rm -rf ./workflowfetched/XXXX/*` in one fast operation?

Ideally, the REST API endpoint should be very fast to reply, so it'll be good not to do any potentially lengthy stuff "synchronously" there, and postpone deletions for overnight and do them "asynchronously". E.g. the REST API endpoint could simply tag this directory as having been submitted e.g. by leaving a file named "LAUNCHED.txt" or something in there, and then the nightly quota updater job would go through all user's `workflowfetched/XXXX` directories and would remove those that were launched already...

Could this "theoretical" concern be "potentially real" based on your scalability tests?

Reply:
If the workflow is successfully launched, then the directory will be empty given that the files are moved from the temporary directory to the workflow's workspace (see [here](https://github.com/reanahub/reana-server/blob/4e8546726881ee6f525434e4089fc99d72d9967c/reana_server/rest/launch.py#L167)). In this case, we are already listing the contents of the directory and moving each entry to the workflow's workspace (see [here](https://github.com/reanahub/reana-server/blob/4e8546726881ee6f525434e4089fc99d72d9967c/reana_server/utils.py#L135)).

Should I investigate whether listing the contents of a directory can be a problem or should we open an issue about it?

Reply:
Since we are going to work on sandboxing soon, I think we can try in the current format and see in practice whether the shared disk slowness wouldn't be a noticeable problem, and come back to this directory tagging etc if necessary.